contrib/flo-styx/src/main/java/com/spotify/flo/contrib/styx/LabelUtil.java

```diff
@@ -0,0 +1,129 @@
+    throw new UnsupportedOperationException();
+  }
+
+  static Map<String, String> buildLabels(String labelPrefix, Config config) {
+    final Map<String, String> map = new HashMap<>();
+
+    final String componentIdLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_COMPONENT_ID_LABEL_KEY);
+    final String workflowIdLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_WORKFLOW_ID_LABEL_KEY);
+    final String parameterLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_PARAMETER_LABEL_KEY);
+    final String executionIdLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_EXECUTION_ID_LABEL_KEY);
+    final String triggerIdLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_TRIGGER_ID_LABEL_KEY);
+    final String triggerTypeLabelKey = validateAndPrefixLabelKey(labelPrefix,
+        STYX_TRIGGER_TYPE_LABEL_KEY);
+    
+    sanitizeLabelValue("c", STYX_COMPONENT_ID, config)
+        .ifPresent(value -> map.put(componentIdLabelKey, value));
+    sanitizeLabelValue("w", STYX_WORKFLOW_ID, config)
+        .ifPresent(value -> map.put(workflowIdLabelKey, value));
+    sanitizeLabelValue("p", STYX_PARAMETER, config)
+        .ifPresent(value -> map.put(parameterLabelKey, value));
+    sanitizeLabelValue("e", STYX_EXECUTION_ID, config)
+        .ifPresent(value -> map.put(executionIdLabelKey, value));
+    sanitizeLabelValue("t", STYX_TRIGGER_ID, config)
+        .ifPresent(value -> map.put(triggerIdLabelKey, value));
+    sanitizeLabelValue("tt", STYX_TRIGGER_TYPE, config)
+        .ifPresent(value -> map.put(triggerTypeLabelKey, value));
+
+    return map;
+  }
+
+  private static String validateAndPrefixLabelKey(String labelPrefix, String label) {
+    final String prefixedLabelKey = String.format("%s-%s", labelPrefix, label);
+
+    if (prefixedLabelKey.length() > LABEL_KEY_MAX_LENGTH) {
+      throw new IllegalArgumentException("Invalid label key: Too long, must be <= 63 characters: "
+                                         + prefixedLabelKey);
+    }
+    
+    if (!prefixedLabelKey.matches(LABEL_KEY_REGEX)) {
+      throw new IllegalArgumentException("Invalid label key: " + prefixedLabelKey);
+    }
+    
+    return prefixedLabelKey;
+  }
+
+  private static Optional<String> sanitizeLabelValue(String pad, String path, Config config) {
+    if (!config.hasPath(path)) {
+      return Optional.empty();
+    }
+
+    // Only allow [-a-z0-9]
+    String value = config.getString(path).toLowerCase().replaceAll(LABEL_VALUE_REPLACE_REGEX, "-");
+    // First character must be [a-z]
+    if (!Character.isAlphabetic(value.charAt(0))) {
+      value = String.format("%s-%s", pad, value);
+    }
+    // Last character must be [a-z0-9] (not -)
+    if (value.charAt(value.length() - 1) == '-') {
+      value = String.format("%s%s", value, pad);
+    }
+    // Trim string if it is too long
+    if (value.length() > LABEL_VALUE_MAX_LENGTH) {
+      value = value.substring(0, LABEL_VALUE_MAX_LENGTH);
```

Comment:
Seems like this can cause `value` to end up being an invalid string even after the applied pad suffix. Why don't we check and cut the length of the label before applying the pad? We can check for a shorter length - seems like 58 would be a good length - that would allow for the prefix and suffix (ex `tt-`... and ..`tt` to later be applied and the label to still be within the limit of 63.

Reply:
Sure we can do that. Just copied what we did in internal Luigi code.